File: iosApp/iosApp/DuplicateDetector.swift
import Foundation

enum DuplicateDetector {
    private static let threshold = 0.75

    static func findDuplicate(newDoc: ScannedDocument, in documents: [ScannedDocument]) -> String? {
        guard !newDoc.ocrText.isEmpty else { return nil }
        let newTokens = tokenize(newDoc.ocrText)
        guard !newTokens.isEmpty else { return nil }

        var bestId: String?
        var bestScore = threshold

        for doc in documents {
            guard doc.id != newDoc.id, !doc.ocrText.isEmpty else { continue }
            let score = jaccard(lhs: newTokens, rhs: tokenize(doc.ocrText))
            if score > bestScore {
                bestScore = score
                bestId = doc.id
            }
        }
        return bestId
    }

    private static func tokenize(_ text: String) -> Set<String> {
        Set(text.lowercased()
            .components(separatedBy: .whitespacesAndNewlines)
            .filter { $0.count > 2 })
    }

    private static func jaccard(lhs: Set<String>, rhs: Set<String>) -> Double {
        guard !lhs.isEmpty || !rhs.isEmpty else { return 0 }
        let intersection = lhs.intersection(rhs).count
        let union = lhs.union(rhs).count
        return union == 0 ? 0 : Double(intersection) / Double(union)
    }
}

File: composeApp/src/commonMain/kotlin/com/mohamedfaridelsherbini/nexar/domain/classifier/DuplicateDetector.kt
package com.mohamedfaridelsherbini.nexar.domain.classifier

import com.mohamedfaridelsherbini.nexar.domain.model.ScannedDocument

object DuplicateDetector : DuplicateDetectionService {

    private const val SIMILARITY_THRESHOLD = 0.75

    private fun tokenize(text: String): Set<String> =
        text.lowercase()
            .split(Regex("""\s+"""))
            .filter { it.length > 2 }
            .toSet()

    private fun jaccard(a: Set<String>, b: Set<String>): Double {
        if (a.isEmpty() && b.isEmpty()) return 0.0
        val intersection = a.intersect(b).size
        val union = a.union(b).size
        return if (union == 0) 0.0 else intersection.toDouble() / union.toDouble()
    }

    override fun findDuplicate(newDoc: ScannedDocument, existingDocs: List<ScannedDocument>): String? {
        if (newDoc.ocrText.isBlank()) return null
        val newTokens = tokenize(newDoc.ocrText)
        if (newTokens.isEmpty()) return null

        var bestId: String? = null
        var bestScore = SIMILARITY_THRESHOLD

        for (doc in existingDocs) {
            if (doc.id == newDoc.id || doc.ocrText.isBlank()) continue
            val score = jaccard(newTokens, tokenize(doc.ocrText))
            if (score > bestScore) {
                bestScore = score
                bestId = doc.id
            }
        }
        return bestId
    }
}
